File: golang/42.go
package leetcode

func trap(height []int) int {
	n, ans := len(height), 0
	l, r := 0, n-1
	hl, hr := height[l], height[r]
	for l < r-1 {
		if hl <= hr {
			l++
			if height[l] < hl {
				ans += hl - height[l]
			} else {
				hl = height[l]
			}
		} else {
			r--
			if height[r] < hr {
				ans += hr - height[r]
			} else {
				hr = height[r]
			}
		}
	}
	return ans
}

File: Rust/src/_42.rs
use crate::leetcode::Solution;

impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let n = height.len();
        let mut ans = 0;
        let (mut l, mut r) = (0 as usize, n - 1);
        let (mut hl, mut hr) = (height[l], height[r]);
        while l < r {
            if hl < hr {
                l += 1;
                if height[l] < hl {
                    ans += hl - height[l];
                } else {
                    hl = height[l];
                }
            } else {
                r -= 1;
                if height[r] < hr {
                    ans += hr - height[r];
                } else {
                    hr = height[r];
                }
            }
        }
        ans
    }
}
